stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -793,6 +793,33 @@ function dot(x::AbstractArray, y::AbstractArray)
     s
 end
 
+"""
+    dot(x, A, y)
+
+Compute the generalized dot product `x' * A * y` between two vectors `x` and `y`.
+For complex vectors, the first vector is conjugated.
+"""
+dot(x::Number, A::Number, y::Number) = conj(x) * A * y
+
+function dot(x::AbstractVector, A::AbstractMatrix, y::AbstractVector)
+    (axes(x)..., axes(y)...) == axes(A) || throw(DimensionMismatch())
+    T = typeof(dot(first(x), first(A), first(y)))
+    zeroxA = zero(adjoint(first(x))*first(A))
```

Comment:
Something like 
```
T = promote_op(dot, eltype(a), eltype(A), eltype(b))
```
and similar for `zeroxA` might be better since that gets inferred (the current formulation isn't, at least on v1.1.1; appears to incur some array-size testing so far as I can tell from `@code_warntype`).

Reply:
Yes, I know. These two lines are written such that the code can handle block matrices/vectors, so eltypes like `Vector` or `Matrix`. The issue is that calling `one` or `zero` on them fails, because the type does not contain any information on the size. And also, the same "`first`" approach is taken in the
```julia
function dot(x::AbstractArray, y::AbstractArray)
```
method a few lines above. Of course, this approach still breaks down when we have blocks of varying lengths, so this may not be a good idea either. To handle that, we would need to initialize `temp` in every loop separately. I'll push a suggestion for how to fix that.